Audit dependencies

brenthosie opened the pull request:
<!-- Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. -->

<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

1. Modernizes the dependencies of this project
2. Ensures linting and tests during commit, push, pr, and publish.
3. Gets the integration tests working after filling out a .env file to contain clientId/clientSecret. Closes #72.
4. Gets the github actions build chain up-to-date

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

https://jira.corp.adobe.com/browse/PDCL-14240

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

Turbine is going to use this sdk to create some libraries and build them. I wanted to ensure that this works with modern node tooling.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

- Ensured all unit tests were passing
- Got the integration tests running
- **Added a test to ensure that commonjs and esm exports have a default function called Reactor**
- **Added jsDom unit-test that loads the built browser bundle and ensures window.Reactor is defined**
- **The commonjs and esm versions are run in jasmine in parallel**
- **Added PlayWright which loads the Reactor SDK with a real access token and calls for companies**

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
pdcl-14240: upgrade @babel/cli, @babel/core, @babel/preset-env, express, node-fetch, nodemon, parcel, semver. force node lts/jod to run the project.

Commit:
pdcl-14240: feat: add dual package support for CommonJS and ES modules

- Add lib/cjs/ CommonJS build alongside existing ES module build
- Configure conditional exports in package.json for require/import compatibility
- Update Babel config with environment-specific module compilation
- Updated elint, glob, mkdirp, prettier, rimraf

Commit:
pdcl-14240: unit and integration tests working.
- fix dev.yaml and npm-publish.yaml.
- v2.0.0-beta.1.

Commit:
pdcl-14240: - update notify script
- v2.0.0-beta.2

Commit:
pdcl-14240: fix the build to be minified for browsers. remove the scripts from served files to npm

Commit:
pdcl-14240: v2.0.0-beta.3

Commit:
v2.0.0-beta.4

Commit:
pdcl-14240: change the restore-dist-tag script to not place the bot token in a file on disk

Commit:
pdcl-14240: clean up files for review

markhicken commented on .github/workflows/dev.yaml:
```diff
@@ -6,22 +6,52 @@ name: ci
+          RSDK_ADOBE_ORG_ID=${{ secrets.RSDK_ADOBE_ORG_ID }}
+          RSDK_ADOBE_REACTOR_COMPANY_ID=${{ secrets.RSDK_ADOBE_REACTOR_COMPANY_ID }}
+          RSDK_ADOBE_REACTOR_URL=${{ secrets.RSDK_ADOBE_REACTOR_URL }}
+          RSDK_ADOBE_ENVIRONMENT=${{ secrets.RSDK_ADOBE_ENVIRONMENT }}
```
Company id, org id, etc are only known to the person who setup the secrets. If something breaks, it could be difficult to troubleshoot. I'm wondering if it might make sense to store these in adobe's internal Vault. We could document the vault path in this workflow file and it would serve as the primary source of truth. We might even consider writing a script that pulls from vault to github via the API.

markhicken replied:
Document any company specific requirements in the .env-example file for this project AND turbine

markhicken commented on scripts/run-integration-tests.js:
```diff
@@ -0,0 +1,195 @@
+  // Don't await in exit handler, just run sync cleanup or ignore
+  if (!cleanupRan) {
+    console.log(`Process exiting with code ${code}, running cleanup...`);
+    await runDeleteTestProperties();
```
Does this actually work? Can you await async things in the exit? Either way, I think a better practice would be to handle cleanup on startup instead of exit. This ensures a clean environment when running tests. It also allows you to inspect the state of things if tests failed.

markhicken commented on package.json:
```diff
@@ -21,101 +21,112 @@
+      "distDir": "dist/",
+      "outputFormat": "global",
+      "isLibrary": true,
+      "sourceMap": false,
```
Any reason not to turn on sourceMap? Our old build had it.

markhicken commented on .env.example:
```diff
@@ -0,0 +1,28 @@
+# prod | stage
+RSDK_ADOBE_ENVIRONMENT=stage
+
+# Note: ACCESS_TOKEN will be generated automatically using @adobe/auth-token
```
Might be helpful to explain this a bit more. When does it get generated? Because the readme says you have to get it yourself. This will likely be confusing for contributors.

markhicken commented on test/integration/helpers.js:
```diff
@@ -390,6 +392,95 @@ const helpers = {
+
+      // Find all properties with "Reactor SDK" prefix
+      await helpers.forEachEntityInList(
+        (paging) => reactor.listPropertiesForCompany(helpers.companyId, paging),
```
This could be filtered instead of listing all properties.

markhicken commented on test/integration/helpers.js:
```diff
@@ -390,6 +392,95 @@ const helpers = {
+        (paging) => reactor.listPropertiesForCompany(helpers.companyId, paging),
+        (property) => {
+          const propName = property.attributes.name;
+          if (propName.startsWith('Reactor SDK ')) {
```
The old cleanup function used the regex so it caught all of them instead of just this one naming scheme. See `deleteTestProperty` which seems to not be in use anymore, but is still in the code.

markhicken replied:
There are a few other naming schemes for properties that we need to update.

brenthosie replied:
I don't think that's the accurate anymore with this being in the PR:

https://github.com/adobe/reactor-sdk-javascript/blob/audit-dependencies/test/integration/helpers.js#L668

markhicken commented on test/integration/helpers.js:
```diff
@@ -390,6 +392,95 @@ const helpers = {
+  // Setup afterAll hook that runs cleanup even if tests fail
+  setupReactorSDKCleanup() {
+    // Global afterAll that runs after all tests in a suite complete
+    afterAll(async function () {
```
Here again, this might be better to run before tests.

markhicken commented on package.json:
```diff
@@ -21,101 +21,112 @@
+    "lint:check": "eslint 'src/**/*.js' 'test/**/*.js' 'scripts/**/*.js'",
+    "lint:staged": "lint-staged",
+    "clean": "rm -rf dist lib && npm run clean:integration:setup",
+    "clean:integration:setup": "rm -rf tmp.tests",
```
We should either use `rimraf` for cross-env compatibility or remove it from the project.

markhicken commented on .babelrc.cjs:
```diff
@@ -0,0 +1,67 @@
+  api.cache(true);
+
+  // Skip Babel config if Parcel is running (Parcel has its own transpilation)
+  if (process.env.PARCEL_WORKER_ID || process.env.NODE_ENV === 'test') {
```
This seems a bit sketchy. Does this mean that our transpilation might be different for tests than for the prod build?

brenthosie replied:
no, not at all. parcel is used to create the browser.bundle and places the output in dist/reactor-sdk.min.js, and babel transpiles the code into lib/*

markhicken commented on scripts/run-integration-tests.js:
```diff
@@ -0,0 +1,195 @@
+
+    // Step 3: Wait for server to be ready
+    console.log('⏳ Waiting for server to start...');
+    await new Promise((resolve) => setTimeout(resolve, 3000));
```
3000 seems like a magic number that won't always be reliable in slower envs. Maybe we should make a waitForServer function?

```javascript
const waitForServer = async (url, maxAttempts = 10, delay = 1000) => {
  for (let i = 0; i < maxAttempts; i++) {
    try {
      await fetch(url);
      return true;
    } catch {
      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }
  throw new Error('Server failed to start');
};

// Replace the fixed sleep with:
await waitForServer('http://localhost:5000');
```

markhicken commented on scripts/run-integration-tests.js:
```diff
@@ -0,0 +1,195 @@
+    );
+    const { stdout } = await execAsync(`node ${scriptPath}`);
+    if (stdout) console.log(stdout.trim());
+    console.log('✅ Cleanup completed');
```
FWIW - All the emojis in this file won't work if run in an environment without full Unicode support.

markhicken commented on scripts/run-integration-tests.js:
```diff
@@ -0,0 +1,195 @@
+
+    // Step 4: Open browsers (only if not in CI environment)
+    if (!process.env.CI && !process.env.HEADLESS) {
+      console.log('🌐 Opening browsers...');
```
`test:integration` sets HEADLESS to true. How would you ever get inside this code block? Would it be worth adding an npm script for headed tests (during local dev)?

Commit:
pdcl-14240: - restore "scripts" as a part of the built files
- v2.0.0-beta.5

markhicken commented on docs/INTEGRATION_TESTING.md:
```diff
@@ -0,0 +1,67 @@
+### 2. Get Your Credentials
+
+Refer to your tech account integration for your Client ID and Client Secret. You can get your company by logging into the
+the Adobe Tags Launch UI and navigating to Data Collection.
```
Not everyone will know what a "tech account" is. Not everyone will have one setup. Can we allow the user to just provide an access token in this case? Maybe if ACCESS_TOKEN exists as an env var we just skip `generateAccessToken`?

brenthosie replied:
I haven't changed how this works. people are free to make a `new Reactor("TOKEN_HERE")` as before.

Edited to add: this .md file is instructions for developers (mainly our team), or external contributors who should know what a tech account is to make updates here.

markhicken commented on src/index.js:
```diff
@@ -12,22 +12,22 @@ governing permissions and limitations under the License.
 
 import fetch from 'node-fetch';
```
Swap to native `fetch`?

markhicken commented on package.json:
```diff
@@ -21,101 +21,113 @@
+    ]
+  },
+  "engines": {
+    "node": ">=v22.19.0"
```
If people are using this dep in their CI this might break stuff. I know we talked about this recently but I'm thinking this might be a bit too restrictive for this package.

brenthosie replied:
They can use the current deployed version until node 20 EOL in April 2026, at which point they should be following the node 22 suggestion here.

markhicken commented on test/integration/generate-access-token.cjs:
```diff
@@ -0,0 +1,70 @@
+
+  const content = `RSDK_ACCESS_TOKEN=${accessToken}\n`;
+
+  fs.writeFileSync(accessTokenEnvPath, content, 'utf-8');
```
Is there any way we can avoid writing it to a file? If not, we should consider changing the permission on the file to `0600` owner access only and immediately cleanup after usage.

brenthosie replied:
resolved. it now exports as a environment variable. also removed writing a .env file in Github Actions.

markhicken commented on src/index.js:
```diff
@@ -198,6 +196,9 @@ Object.assign(
+// Add version property to the prototype
+Reactor.prototype.version = version;
+
+if (typeof window !== 'undefined') {
```
This could cause issues in non-browser environments that define window (e.g., some server-side rendering contexts). Claude suggested reverting this line. I'm not sure if this change was required based on the other testing changes.

brenthosie replied:
wouldn't server-side rendering use the module or commonjs versions?

markhicken commented on scripts/fix-cjs-paths.cjs:
```diff
@@ -0,0 +1,47 @@
+  const updatedContent = content.replace(
+    /require\(["'](\.[^"']*?)\.js["']\)/g,
+    'require("$1.cjs")'
+  );
```
This regex doesn't handle edge cases like:
- Requires with backticks
- Dynamic requires with template strings
- Requires split across lines

brenthosie replied:
Can you demonstrate the problem you're perceiving?

markhicken commented on docs/INTEGRATION_TESTING.md:
```diff
@@ -0,0 +1,67 @@
+# Integration Testing Setup
```
Can we link to this file from the `README.md`?

markhicken commented on docs/INTEGRATION_TESTING.md:
````diff
@@ -0,0 +1,67 @@
+### 3. Configure Your .env File
+
+```bash
+RSDK_ADOBE_CLIENT_ID=your_client_id
````
Is there a need for such tightly scoped ENV var names? The `RSDK_ADOBE_` doesn't seem totally necessary. If you think we need scoping, maybe we could just use `REACTOR_`?

markhicken commented on scripts/fix-cjs-paths.cjs:
```diff
@@ -0,0 +1,47 @@
+when using --out-file-extension .cjs
+*/
+
+const fs = require('fs');
```
nitpick - Because this is run after the build, there is no testing to validate that it does what we expect, right? I wish we could just do this with a babel plugin or something.

brenthosie replied:
i have tests that look at the commonjs and module exports. it should cover this file behaving as expected.

brenthosie replied:
I now run the commonjs and esm bundles in jasmine against the full test suite.

markhicken commented on package.json:
```diff
@@ -21,101 +21,113 @@
-  "browser": "dist/reactor-sdk-library.min.js",
-  "module": "dist/reactor-sdk-bundled.min.js",
-  "main": "./lib/node/index.js",
+  "type": "module",
```
I created a new project, installed `2.0.0-beta.5` and made a very basic node script that tries to list companies. When running my script, I get this warning...

```
$ node index.js
(node:97552) [MODULE_TYPELESS_PACKAGE_JSON] Warning: Module type of file:///index.js is not specified and it doesn't parse as CommonJS.
Reparsing as ES module because module syntax was detected. This incurs a performance overhead.
To eliminate this warning, add "type": "module" to package.json.
```

I find this very strange because `"type": "module"` is in my `package.json` in my `node_modules` folder. I wonder if this has something to do with the fact that we have a `main`, `module`, `browser` and `exports`. Some of these override others, right?

brenthosie replied:
I didn't encounter this problem with a minimal example for myself. We'll probably have to debug over a call to see your error.

markhicken commented on README.md:
Using this CDN URL directly in the browser causes a CORS error.
`<script src="https://unpkg.com/@adobe/reactor-sdk@2.0.0-beta.5/dist/reactor-sdk.min.js"></script>`

```
(index):1 Access to fetch at 'https://reactor.adobe.io/companies' from origin 'http://127.0.0.1:8080' has been blocked by CORS policy: Request header field cache-control is not allowed by Access-Control-Allow-Headers in preflight response.
reactor-sdk.min.js:1  GET https://reactor.adobe.io/companies net::ERR_FAILED
```

If I manually delete the `cache-control` header before making a request, it works.
`delete reactor.headers['Cache-control'];`

We probably need to strip that header for browsers.

markhicken reviewed (changes requested):
Nice update! Great work @brenthosie. I've added some comments.

Commit:
pdcl-14240: add "if: always()" during cleaning of sensitive files in dev.yaml and npm-publish.yaml

Commit:
pdcl-14240: address PR concerns

Commit:
2.0.0-beta.6

Commit:
pdcl-14240: document company requirements to run tests. fix issue with build:production race condition.

Commit:
pdcl-14240: using NODE_ENV=test isn't working now that i've fixed how the test prompts work.

Commit:
v2.0.0-beta.7

Commit:
pdcl-14240: Uninstall nock. Begin cleaning up how the integration tests import and run. Remove access token as a file.

Commit:
don't enable logging by default

Commit:
preparing to also run commonjs tests

Commit:
pdcl-14240: Able to build the test suite for commonjs bundle.

Commit:
pdcl-14240: run the integration tests in parallel for commonjs and esmodule

Commit:
pdcl-14240: delay reporting of test suites together

Commit:
pdcl-14240: ensure that credentials are always read from process.env during CI

Commit:
pdcl-14240: wrap the actual results with the synchronized test results box

Commit:
pdcl-14240: clean up the test runner script. gracefully shut down the express server.

Commit:
pdcl-14240: ensure README.md and INTEGRATION_TESTING.md are consistent for integration tests.

Commit:
pdcl-14240: ensure parallel runs avoid naming collisions for resources.

Commit:
pdcl-14240: use playwright for a browser build sanity check

Commit:
v2.0.0-beta.8

Commit:
pdcl-14240: update INTEGRATION_TESTING.md instructions

Commit:
pdcl-14240: skip the playwright post-install browser download so we can use caching and ensure browsers after

Commit:
v2.0.0-beta.9

Commit:
pdcl-14240: 2.0.0-beta.10 - remove cache from ci flow

Commit:
pdcl-14240: v2.0.0-beta.11 - forgot to remove playwright browser install in dev.yaml ci

Commit:
pdcl-14240: make the commonjs and esm runs sequential in case failure are due to rate limiting.

Remove unnecessary express server for tests.

Clean up run-integration-tests file.

Commit:
v2.0.0-beta.12

jhicken commented on test/integration/reactor.js:
```diff
@@ -17,7 +17,7 @@ var reactor = globals.reactor;
   const options = {
     reactorUrl: globals.REACTOR_URL,
-    enableLogging: true,
+    enableLogging: Boolean(globals.enableLogging),
```
Maybe I missed it. Is there a comment on how to turn this on somewhere?

jhicken commented on test/integration/helpers.js:
```diff
@@ -568,19 +659,13 @@ function makeNameForTestObject(objectType, baseName) {
-  }
-  return testingProperties.get(entityKind);
+  // Examples:
+  // Reactor SDK AuditEvent-Testing Base (Integration Testing Property / 2025-10-09T14:23:17.456-07:00) 1a2b3c4d5e6f7890
```
thats super helpful!

jhicken commented on scripts/run-integration-tests.js:
```diff
@@ -0,0 +1,328 @@
+          testProcess = spawn('jasmine', [bundlePath], {
+            stdio: ['inherit', 'pipe', 'pipe']
+          });
+          console.log(`Starting ${testType} tests...`);
```
this is a little nitpick. maybe it would make more sense to say we are starting before we spawn the process. Then if there is an explosion in spawning we would have this log point before the explosion.

jhicken reviewed (approved):
I did not have much to add. There was a ton of non functionality changes that made this gnarly.

Looks good to me.
